src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java

```diff
@@ -0,0 +1,499 @@
+     * stop sequence to avoid overwriting these other stop times.
+     */
+    private static void insertBlankStopTimes(
+        String tablePrefix,
+        List<String> tripIds,
+        List<Pattern> genericStops,
+        int startingStopSequence,
+        int stopTimesToAdd,
+        Connection connection
+    ) throws SQLException {
+        if (tripIds.isEmpty()) {
+            // There is no need to insert blank stop times if there are no trips for the pattern.
+            return;
+        }
+        String insertSql = Table.STOP_TIMES.generateInsertSql(tablePrefix, true);
+        PreparedStatement insertStatement = connection.prepareStatement(insertSql);
+        int totalRowsUpdated = 0;
+        // Create a new stop time for each sequence value (times each trip ID) that needs to be inserted.
+        for (int i = startingStopSequence; i < stopTimesToAdd + startingStopSequence; i++) {
+            StopTime stopTime = genericStops.get(i).stopTime;
+            stopTime.stop_sequence = i;
+            // Update stop time with each trip ID and add to batch.
+            for (String tripId : tripIds) {
+                stopTime.trip_id = tripId;
+                stopTime.setStatementParameters(insertStatement, true);
+                insertStatement.addBatch();
+                int[] rowsUpdated = insertStatement.executeBatch();
+                totalRowsUpdated += rowsUpdated.length;
+            }
+        }
+        int[] rowsUpdated = insertStatement.executeBatch();
+        totalRowsUpdated += rowsUpdated.length;
+        LOG.info("{} blank stop times inserted", totalRowsUpdated);
+    }
+
+    /**
+     * Check that the stops in the changed region remain in the same order. If not, throw an exception to cancel the
+     * transaction.
+     */
+    private static void verifyInteriorStopsAreUnchanged(
+        List<String> originalStopIds,
+        List<String> newStopIds,
+        int firstDifferentIndex,
+        int lastDifferentIndex
+    ) {
+        for (int i = firstDifferentIndex; i <= (lastDifferentIndex - 1); i++) {
+            String newStopId = newStopIds.get(i);
+            // Because a stop was inserted at position firstDifferentIndex, all original stop ids are shifted by one.
+            String originalStopId = originalStopIds.get(i + 1);
+            if (!newStopId.equals(originalStopId)) {
+                // If the new stop ID at the given index does not match the original stop ID, the order of at least
+                // one stop within the changed region has been changed. This is illegal according to the rule enforcing
+                // only a single addition, deletion, or transposition per update.
+                throw new IllegalStateException(RECONCILE_STOPS_ERROR_MSG);
+            }
+        }
+    }
+
+    /**
+     * Generic pattern class use to hold either a pattern stop or pattern location derived data.
+     */
+    private static class Pattern {
```

Comment:
Maybe this didn't make the previous round. Rename this class to `GenericStop` or something.